### ISSUE TITLE
Bi directional I/O?

### ISSUE THREAD

**Issue by Mekelaina** [hneemann/Digital#1307]
Love the program but I've found implementing a data bus like in older computers to be difficult. The 6502 for example, had 8 data lines with a pin for whether the CPU was reading or writing to those lines. And I've not found a decent solution other than making data in pins and data out pins. Which is fine for simulations but makes prototyping physical circuits frustrating.

**Comment by hneemann** [comment]
Each pin is bi-directional by default. In fact, there are no uni-directional pins in this simulator. 
Making a pin an input or output only affects the shape of the pin and how you can interact with that shape.

**Comment by shawty** [comment]
would it be possible to create a sample showing us how we might construct a bi-directional bus controlled by another signal such as those used for r/w signals?   I to have found trying to do this very frustrating.

**Comment by hneemann** [comment]
It works just like a real bus. If open-collector drivers are used, you can connect all wires together; if TriState outputs are used, you need a control circuit that ensures that only one output is active. An input can be connected to the bus for reading.

Inside the sub-circuit it would look something like this:
![z](https://github.com/user-attachments/assets/fa5bbafb-1245-4f11-8cbd-60d5f71f2da9)

**Comment by hneemann** [comment]
And if an open-collector bus is to be used, it looks like this:

![z](https://github.com/user-attachments/assets/c0019b37-b1b6-44f1-9eb0-845e9f27b4cc)

If you want to read, /out must be set to zero. And there must be a pull-up somewhere on the bus.

**Comment by shawty** [comment]
Can we do "pull ups" in Digital?  That's an analog function right?  or would it suffice to just connect a line to a +ve terminal?

**Comment by hneemann** [comment]
If you connect a wire to a +vcc terminal, you can never pull it down. Therefore the pull-up is necessary.
Whether a pull-up is connected to a wire or not is actually just a flag. If it is set, the wire is set to high when all connected outputs are in the high-z state. It is not really an analog function.

**Comment by shawty** [comment]
Going with your sub circuit above (First Image) this is what I've basically been trying to do:

![image](https://github.com/user-attachments/assets/b81876cf-c52d-4fb0-a9fd-82cadf453e9b)

With a sub circuit inside the IO-Reg:

![image](https://github.com/user-attachments/assets/c2ce80de-d152-41bf-af38-b291c8507188)

![image](https://github.com/user-attachments/assets/dbaa12bd-d978-457e-a78e-5492b18d4af3)

![image](https://github.com/user-attachments/assets/6f52e0c4-9881-4c32-b381-f50346bf3dbf)

The I/O's etc on the external Circuit look like the following:

![image](https://github.com/user-attachments/assets/646bce8d-c94c-4800-ad43-21f466708270)

![image](https://github.com/user-attachments/assets/61279007-a302-4b2b-8f76-ed27d5f53513)

![image](https://github.com/user-attachments/assets/50370fef-e864-4bd7-bfcc-e909fe78b03d)

![image](https://github.com/user-attachments/assets/039a444a-5a7c-49c2-b7cf-84bd8da6a5f0)

I think, @Mekelaina was potentially trying to do something similar to me.  In my efforts to model a 6522 VIA for example, the device has 16 r/w registers connected directly to a common data bus.  the register to access is controlled by 4 Address pins, and r/w is controlled by a regular R not W line, where it's High to read the internal register, and Low to Write to it.

**Comment by shawty** [comment]
Everything works up to the point where I try to set the IO-Reg to be writable, and I get an output collision.

![image](https://github.com/user-attachments/assets/ffad2908-dc1a-43c4-95fe-f987efa2a848)

**Comment by hneemann** [comment]
Remove the data input or set its default value to “high z”. 
If the input component outputs a value and the register also outputs a value, this leads to a short circuit.

[bus.zip](https://github.com/user-attachments/files/16752538/bus.zip)

**Comment by shawty** [comment]
Ah I see what your doing now, ok, I'll have a play with that and see what transpires.  I did at one point even try to write this as a Java Component too :-)

For reference, this is what I've been trying to emulate in Digital (For quite some time now) :-D

http://6502.org/documents/datasheets/rockwell/rockwell_r6522_via.pdf

**Comment by KubaO** [comment]
> Can we do "pull ups" in Digital? That's an analog function right? or would it suffice to just connect a line to a +ve terminal?

Yes. See Wires->Pull-Up Resistor. They set the signal high if it's not driven. There is also a pull-down resistor.
